haystack/components/agents/agent.py

```diff
@@ -205,48 +233,59 @@ def run(
         if not self._is_warmed_up and hasattr(self.chat_generator, "warm_up"):
             raise RuntimeError("The component Agent wasn't warmed up. Run 'warm_up()' before calling 'run()'.")
 
-        state = State(schema=self.state_schema, data=kwargs)
-
         if self.system_prompt is not None:
             messages = [ChatMessage.from_system(self.system_prompt)] + messages
-        state.set("messages", messages)
-
-        generator_inputs: Dict[str, Any] = {"tools": self.tools}
-
-        selected_callback = streaming_callback or self.streaming_callback
-        if selected_callback is not None:
-            generator_inputs["streaming_callback"] = selected_callback
 
-        # Repeat until the exit condition is met
-        counter = 0
-        while counter < self.max_agent_steps:
-            # 1. Call the ChatGenerator
-            llm_messages = self.chat_generator.run(messages=messages, **generator_inputs)["replies"]
-            state.set("messages", llm_messages)
+        input_data = deepcopy({"messages": messages, "streaming_callback": streaming_callback, **kwargs})
 
-            # 2. Check if any of the LLM responses contain a tool call
-            if not any(msg.tool_call for msg in llm_messages):
-                return {**state.data}
+        state = State(schema=self.state_schema, data=kwargs)
+        state.set("messages", messages)
 
-            # 3. Call the ToolInvoker
-            # We only send the messages from the LLM to the tool invoker
-            tool_invoker_result = self._tool_invoker.run(messages=llm_messages, state=state)
-            tool_messages = tool_invoker_result["tool_messages"]
-            state = tool_invoker_result["state"]
-            state.set("messages", tool_messages)
+        generator_inputs = self._prepare_generator_inputs(streaming_callback=streaming_callback)
+
+        component_visits = dict.fromkeys(["chat_generator", "tool_invoker"], 0)
+        with self._create_agent_span(input_data) as span:
+            counter = 0
+            while counter < self.max_agent_steps:
+                # 1. Call the ChatGenerator
+                llm_messages = Pipeline._run_component(
+                    component_name="chat_generator",
+                    component={"instance": self.chat_generator},
+                    inputs={"messages": messages, **generator_inputs},
+                    component_visits=component_visits,
+                    parent_span=span,
+                )["replies"]
+                state.set("messages", llm_messages)
+
+                # 2. Check if any of the LLM responses contain a tool call
+                if not any(msg.tool_call for msg in llm_messages):
+                    return self._finalize_run(state=state, span=span, counter=counter)
```

Comment:
It stands out for me that we call `_finalize_run` three times when perhaps we could call it once in finally or some similar exit point? And also perhaps `_finalize_run` should not return state, it kinda hides/indirects what's going on and what is returned.

Reply:
I think it’s fine to call three times. But happy to take the state part out and leave that explicit and just have do the span stuff.

Reply:
Gave this more thought and I agree, we can utilize break here for when we are in the loop instead of needing to call the same return three times.